bitwarden/mobile
# Improve Theming

## Type of change

## Objective

Improve theming logic and performance, also fix issues on theme applying on password generator and password generator history. Also fix exception raise on applying theme because of collection modification on enumeration because of the merged dictionaries.

It should fix [https://appcenter.ms/users/kspearrin/apps/bitwarden/crashes/errors/848906559u/overview](https://appcenter.ms/users/kspearrin/apps/bitwarden/crashes/errors/848906559u/overview
)

## Code changes



* **ThemeManager.cs:** Changed several things here to improve performance like checking whether the current theme is already the one we want to set. Also only change the theme ResourceDictionary instead of clearing everything and adding all resources again, this was crashing the app on previous versions and raising exceptions leading to inconsistent theming state on some screens
* **MainActivity.cs and AppDelegate.cs:** Added messenger send to report when the app appeared so that we can update the password generated screens on theme change
* **Generator...:** Added logic to update the theming correctly on these screens
* **App.xaml.cs:** Changed UpdateTheme so that it's done in proper order and applied without any issue on race conditions

## Testing requirements


Steps

- Unlock the app
- Put the app to background
- Change OS theme
- Wait the vault timeout (you can set it previously on the app to 1 minute so you don't have to wait that much)
- Return to the app
- The colors and theme are weird on some places of the lock screen

This also affects Password Generator and Password Generator History screens, i.e.:

- Go to Password Generator screen and then to its history (three dots)
- Change the OS theme without waiting the vault timeout
- Go back to the app without waiting the vault timeout
- You'll see that not all passwords themes are changed
- Also if you go back, the generated password of the Password Generator screen will be with previous theme 

## Before you submit
